src/java.base/share/classes/java/util/Formatter.java

```diff
@@ -2448,13 +2442,17 @@ public Formatter(OutputStream os, Charset charset, Locale l) {
         this(l, new BufferedWriter(new OutputStreamWriter(os, charset)));
     }
 
-    private static char getZero(Locale l) {
-        if ((l != null) && !l.equals(Locale.US)) {
-            DecimalFormatSymbols dfs = DecimalFormatSymbols.getInstance(l);
-            return dfs.getZeroDigit();
-        } else {
-            return '0';
+    private char zero() {
+        char zero = this.zero;
+        if (zero == 0) {
```

Comment:
U+0000 is a valid character. Although it's almost no possibility where any locale assigns it as the zero in it, theoretically it is possible. It can be avoided by comparing it with a non-character, such as BOM (U+FFFE)

Reply:
Ok, maybe a bit too defensive, but let's use BOM instead.